In [1]:
#imports
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt

In [2]:
df = pd.read_excel("data/dataset.xlsx", engine="openpyxl")

In [3]:
df.columns

Index(['Num', 'Timestamp', 'Current_J0', 'Temperature_T0', 'Current_J1',
       'Temperature_J1', 'Current_J2', 'Temperature_J2', 'Current_J3',
       'Temperature_J3', 'Current_J4', 'Temperature_J4', 'Current_J5',
       'Temperature_J5', 'Speed_J0', 'Speed_J1', 'Speed_J2', 'Speed_J3',
       'Speed_J4', 'Speed_J5', 'Tool_current', 'cycle ',
       'Robot_ProtectiveStop', 'grip_lost'],
      dtype='object')

In [4]:
# Drop irrelevant columns
df.drop(["Timestamp", "Num", "cycle ", "Robot_ProtectiveStop"], axis=1, inplace=True)

# Show data
df.head(10)

Current_J0  Temperature_T0  Current_J1  Temperature_J1  Current_J2  \
0    0.109628         27.8750   -2.024669         29.3750   -1.531442   
1    0.595605         27.8750   -2.278456         29.3125   -0.866556   
2   -0.229474         27.8750   -2.800408         29.3125   -2.304336   
3    0.065053         27.8750   -3.687768         29.3125   -1.217652   
4    0.884140         27.8750   -2.938830         29.3750   -1.794076   
5    0.118961         27.8125   -2.162542         29.3750   -1.211779   
6    0.086138         27.8750   -1.757647         29.3750   -0.960634   
7    0.075657         27.8750   -1.836465         29.3750   -0.961322   
8   -0.186848         27.8750   -2.563738         29.3750   -1.418934   
9   -0.158912         27.8750   -1.750777         29.3750   -1.979400   

   Temperature_J2  Current_J3  Temperature_J3  Current_J4  Temperature_J4  \
0         29.3750   -0.998570         32.1250   -0.062540         32.2500   
1         29.4375   -0.206097         32.1875   -1.062762         32.2500   
2         29.4375   -0.351499         32.1250   -0.668869         32.3125   
3         29.4375   -1.209115         32.1250   -0.819755         32.2500   
4         29.4375   -2.356471         32.1875   -0.966427         32.3125   
5         29.4375   -0.481834         32.1875    0.015318         32.3125   
6         29.4375   -0.488380         32.1875    0.006506         32.3125   
7         29.4375   -0.512541         32.1875   -0.021845         32.3125   
8         29.4375   -0.262015         32.1875    0.099942         32.3125   
9         29.4375   -0.184550         32.1875    0.079157         32.3125   

   Current_J5  Temperature_J5      Speed_J0  Speed_J1  Speed_J2  Speed_J3  \
0   -0.152622         32.0000  2.955651e-01 -0.000490  0.001310 -0.132836   
1   -0.260764         32.0000 -7.391485e-30 -0.000304  0.002185  0.001668   
2    0.039071         32.0625  1.369386e-01  0.007795 -2.535874  0.379867   
3    0.153903         32.0000 -9.030032e-02 -0.004911 -0.009096 -0.384196   
4    0.178998         32.0000  1.268088e-01  0.005567  0.001138 -0.353284   
5    0.064844         32.0625 -4.639511e-05  0.000093  0.000262  0.002171   
6    0.053249         32.0625  0.000000e+00  0.000000  0.000000  0.000000   
7    0.039228         32.0625  0.000000e+00  0.000000  0.000000  0.000000   
8   -0.173527         32.0625 -1.660036e-01  0.002409 -0.015483  0.350063   
9   -0.209541         32.0625  9.832304e-02  0.004996 -0.016996  0.496241   

   Speed_J4      Speed_J5  Tool_current  grip_lost  
0 -0.007479 -1.529622e-01      0.082732      False  
1 -0.000767  4.169016e-04      0.505895      False  
2  0.000455 -4.968559e-01      0.079420      False  
3  0.018411  4.255591e-01      0.083325      False  
4  0.014994  1.809886e-01      0.086379      False  
5  0.000265 -2.470161e-03      0.188310      False  
6  0.000000  1.121039e-44      0.085192      False  
7  0.000000  0.000000e+00      0.087058      False  
8 -0.009986 -1.927102e-01      0.085785      False  
9 -0.017353 -5.886860e-01      0.085020      False

In [5]:
# Change grip_lost to binary values instead of boolean falues, only use this cell once otherwise all grip_lost turn to null values
df["grip_lost"] = df["grip_lost"].map({False: 0, True: 1})

In [6]:
# Features and target
y = df["grip_lost"]
X = df.drop(["grip_lost"], axis=1)

class_counts = y.value_counts()
print("Original Class Distribution:")
print(class_counts)

Original Class Distribution:
grip_lost
0    7166
1     243
Name: count, dtype: int64


In [7]:
# train and test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=11)

In [8]:
# normalizing data using numpy

# Mean of features
mean = np.mean(X_train, axis=0)

# std of features
std = np.std(X_train, axis=0) 

# normalize training set
X_train_normalized = (X_train - mean) / std

# Normalize test set using training set parameters
X_test_normalized = (X_test - mean) / std

# reshaping y
y_train = y_train.values.reshape(-1, 1)  # convert to column vector
y_test = y_test.values.reshape(-1, 1)


# Handle potential NaN or Inf values aftr normalization
X_train_normalized = np.nan_to_num(X_train_normalized, nan=0.0, posinf=0.0, neginf=0.0)
X_test_normalized = np.nan_to_num(X_test_normalized, nan=0.0, posinf=0.0, neginf=0.0)
# print("Any NaN in y_train:", np.isnan(y_train).any())
# print("Any Inf in X_train:", np.isinf(X_train).any())
# print("Any Inf in y_train:", np.isinf(y_train).any())

print(X_test_normalized)

[[-1.04739934e-01 -2.24610863e-01 -6.39001885e-01 ...  2.45139426e-02
  -8.62654817e-04 -2.93865080e-01]
 [-6.15802452e-03 -2.52392403e+00  6.71725757e-01 ...  2.45139426e-02
  -8.62654817e-04 -2.22603657e-01]
 [ 2.67703523e-01  4.58353444e-01  6.56081181e-01 ... -1.09364484e-02
   2.02667114e-02  1.80474950e+00]
 ...
 [ 1.70150964e-01 -1.95478711e+00  5.78316098e-01 ...  2.45139426e-02
  -8.62654817e-04 -3.12258257e-01]
 [ 9.45840493e-02  5.72180828e-01  7.07024986e-01 ...  2.45139426e-02
  -8.62654817e-04 -3.75968373e-01]
 [-3.59156944e-01 -1.47671209e+00 -7.33282976e-01 ...  1.85971084e-01
   2.86083296e-01 -3.18714984e-01]]


From here I am going to build a neural network.

In [9]:
# NN config
input_size = X_train.shape[1] # amount of features is the input layer
hidden_size = 16 # random set amount of hidden size 
output_size = 1 # only one output neuron that is binary of course
learning_rate = 1e-3
epochs = 500


In [10]:
np.random.seed(11)
W1 = np.random.randn(input_size, hidden_size) * 0.01
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size) * 0.01
b2 = np.zeros((1, output_size))


In [11]:
# Sigmoid activation function
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

# Sigmoid derivative
def sigmoid_derivative(a):
    return a * (1 - a)


In [12]:
def forward_propagation(X, W1, b1, W2, b2):
    # Input layer to hidden layer
    Z1 = np.dot(X, W1) + b1
    A1 = sigmoid(Z1)
    
    # Hidden layer to output layer
    Z2 = np.dot(A1, W2) + b2
    A2 = sigmoid(Z2)
    A2 = np.clip(A2, 1e-7, 1 - 1e-7)  # Avoid exact 0 or 1 values

    
    return Z1, A1, Z2, A2


In [13]:
def backward_propagation(X, y, Z1, A1, Z2, A2, W2):
    m = X.shape[0] 
    
    # Compute output error
    dZ2 = A2 - y.reshape(-1, 1)
    dW2 = (1 / m) * np.dot(A1.T, dZ2)
    db2 = (1 / m) * np.sum(dZ2, axis=0, keepdims=True)
    
    # Compute hidden layer error
    dA1 = np.dot(dZ2, W2.T)
    dZ1 = dA1 * sigmoid_derivative(A1)
    dW1 = (1 / m) * np.dot(X.T, dZ1)
    db1 = (1 / m) * np.sum(dZ1, axis=0, keepdims=True)
    
    return dW1, db1, dW2, db2


In [14]:
# Update the training loop to use normalized data
for epoch in range(epochs):
    # Forward propagation on normalized training data
    Z1, A1, Z2, A2 = forward_propagation(X_train_normalized, W1, b1, W2, b2)

    # Loss function (BCE)
    epsilon = 1e-9  # Small constant to avoid log(0)
    loss = -np.mean(y_train * np.log(A2 + epsilon) + (1 - y_train) * np.log(1 - A2 + epsilon))

    # Backpropagation
    dW1, db1, dW2, db2 = backward_propagation(X_train_normalized, y_train, Z1, A1, Z2, A2, W2)

    # Gradient Descent (Update Weights)
    W1 -= learning_rate * dW1
    b1 -= learning_rate * db1
    W2 -= learning_rate * dW2
    b2 -= learning_rate * db2

    # Print loss every 50 epochs
    if epoch % 50 == 0:
        print(f"Epoch {epoch}, Loss: {loss:.4f}")


Epoch 0, Loss: 0.6848
Epoch 50, Loss: 0.6350
Epoch 100, Loss: 0.5911
Epoch 150, Loss: 0.5522
Epoch 200, Loss: 0.5178
Epoch 250, Loss: 0.4872
Epoch 300, Loss: 0.4601
Epoch 350, Loss: 0.4358
Epoch 400, Loss: 0.4142
Epoch 450, Loss: 0.3947


In [15]:
# Predict on test data
_, _, _, A2_test = forward_propagation(X_test, W1, b1, W2, b2)
y_pred = (A2_test > 0.5).astype(int)

# Compute accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Test Accuracy: {accuracy:.4f}")


Test Accuracy: 0.9777


Below here is model evaluation here I don't just use numpy and pandas, because too much work and not being that educational.
We can see that the model learned to predict the majority class. The test accuracy seems high 0.9777, however this is achieved by predicting only class 0, no grip_lost. This can probably be fixed, but this wasn't the goal of this project, which was building a NN from scratch using only Numpy and Pandas.

In [16]:
# Import necessary libraries
from sklearn.metrics import classification_report

def forward_propagation(X, W1, b1, W2, b2):
    # Input layer to hidden layer
    Z1 = np.dot(X, W1) + b1
    A1 = 1 / (1 + np.exp(-Z1))  # Sigmoid activation
    
    # Hidden layer to output layer
    Z2 = np.dot(A1, W2) + b2
    A2 = 1 / (1 + np.exp(-Z2))  # Sigmoid activation
    A2 = np.clip(A2, 1e-7, 1 - 1e-7)  # Clip so we don't get exact 1 or 0 values
    
    return Z1, A1, Z2, A2

# Predict on test data using normalized inputs
_, _, _, A2_test = forward_propagation(X_test_normalized, W1, b1, W2, b2)

# Convert probabilities to binary predictions
y_pred = (A2_test > 0.5).astype(int)

# Generate classification report
report = classification_report(
    y_test, 
    y_pred, 
    target_names=['No Grip Lost', 'Grip Lost']
)

# Print the Classification Report
print("Classification Report:\n", report)


Classification Report:
               precision    recall  f1-score   support

No Grip Lost       0.98      1.00      0.99      1449
   Grip Lost       0.00      0.00      0.00        33

    accuracy                           0.98      1482
   macro avg       0.49      0.50      0.49      1482
weighted avg       0.96      0.98      0.97      1482



c:\Users\mpjgl\miniconda3\envs\seg3d\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\mpjgl\miniconda3\envs\seg3d\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Users\mpjgl\miniconda3\envs\seg3d\lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
